urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/farom/astroiddriver/INDIAstroidDriver.java

```diff
@@ -0,0 +1,1636 @@
+				Constants.PropertyStates.IDLE, Constants.PropertyPermissions.RW); // REL_FOCUS_POSITION
+		relFocusPosE = new INDINumberElement(relFocusPosP, "FOCUS_RELATIVE_POSITION", "Rel position", 0, -1e9, 1e9, 0,"%7.2f"); // FOCUS_RELATIVE_POSITION
+		
+		absFocusPosP = new INDINumberProperty(this, "ABS_FOCUS_POSITION", "Absolute position", "Focuser / intervalometer",
+				Constants.PropertyStates.IDLE, Constants.PropertyPermissions.RW); // ABS_FOCUS_POSITION
+		absFocusPosE = new INDINumberElement(absFocusPosP, "FOCUS_ABSOLUTE_POSITION", "Abs position", 0, -1e9, 1e9, 0,"%7.2f"); // FOCUS_ABSOLUTE_POSITION
+		
+		// --- Remaining initializations ---
+
+		lastStatusMessage = new StatusMessage();
+		command = new CmdMessage();
+
+		// --- Setup 1s timer (sidereal time update & link status verification)
+		// ---
+		TimerTask task = new TimerTask() {
+			@Override
+			public void run() {
+				lstE.setValue(getSiderealTime());
+				if (isConnected()) {
+					if ((new Date()).getTime() - lastStatusMessage.time > CONNECTION_TIMEOUT_ALERT) {
+						linkStatusE.setValue(LightStates.ALERT);
+					} else {
+						linkStatusE.setValue(LightStates.OK);
+					}
+
+					try {
+						updateProperty(timeLstP);
+						updateProperty(linkStatusP);
+					} catch (INDIException e) {
+						e.printStackTrace();
+					}
+				}
+			}
+		};
+		Timer timer = new Timer();
+		timer.scheduleAtFixedRate(task, 0, 1000);
+
+
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see laazotea.indi.driver.INDIDriver#getName()
+	 */
+	@Override
+	public String getName() {
+		return "INDI Astroid Driver";
+	}
+
+	/**
+	 * Called when a new BLOB Vector message has been received from a Client.
+	 * 
+	 * @param property
+	 *            The BLOB Property asked to change.
+	 * @param timestamp
+	 *            The timestamp of the received message
+	 * @param elementsAndValues
+	 *            An array of pairs of BLOL Elements and its requested values to
+	 *            be parsed.
+	 */
+	@Override
+	public void processNewBLOBValue(INDIBLOBProperty property, Date date, INDIBLOBElementAndValue[] elementsAndValues) {
+
+	}
+
+	/**
+	 * Called when a new Number Vector message has been received from a Client.
+	 * 
+	 * @param property
+	 *            The Number Property asked to change.
+	 * @param timestamp
+	 *            The timestamp of the received message
+	 * @param elementsAndValues
+	 *            An array of pairs of Number Elements and its requested values
+	 *            to be parsed.
+	 */
+	@Override
+	public void processNewNumberValue(INDINumberProperty property, Date date,
+			INDINumberElementAndValue[] elementsAndValues) {
+
+		
+		
+		try{
+			// Avoid crash when empty property
+			if(elementsAndValues == null){			
+				try {
+					printMessage("elementsAndValues == null");
+					property.setState(PropertyStates.ALERT);
+					updateProperty(property, "Empty property: you may have enter an invalid value");
+				} catch (INDIException e) {
+					e.printStackTrace();
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 388:394, 398:404, 817:823, 827:833, 1152:1158, 1162:1168.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.